apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/ThrowableTransformer.java

```diff
@@ -43,11 +47,12 @@ public boolean overMaxLength() {
                 }
             });
 
-            if (overMaxLength) {
+            if (isLookDeeper) {
                 break;
             }
 
-            causeException = throwable.getCause();
+            causeException = causeException.getCause();
```

Comment:
The other one triggers #2926

Reply:
FYI @ascrutae This is the root cause and being here for a long time. `OverSize` check avoids OOM in most cases only.